language/move-binary-format/src/compatibility.rs

```diff
@@ -2,40 +2,36 @@
 // Copyright (c) The Move Contributors
 // SPDX-License-Identifier: Apache-2.0
 
+use std::collections::BTreeSet;
+
 use crate::{
+    errors::{PartialVMError, PartialVMResult},
     file_format::{AbilitySet, StructTypeParameter, Visibility},
     file_format_common::VERSION_5,
     normalized::Module,
 };
-use std::collections::BTreeSet;
+use move_core_types::vm_status::StatusCode;
+
 
 /// The result of a linking and layout compatibility check. Here is what the different combinations
 /// mean:
-/// `{ struct: true, struct_layout: true }`: fully backward compatible
-/// `{ struct_and_function_linking: true, struct_layout: false }`: Dependent modules that reference functions or types in this module may not link. However, fixing, recompiling, and redeploying all dependent modules will work--no data migration needed.
-/// `{ type_and_function_linking: true, struct_layout: false }`: Attempting to read structs published by this module will now fail at runtime. However, dependent modules will continue to link. Requires data migration, but no changes to dependent modules.
-/// `{ type_and_function_linking: false, struct_layout: false }`: Everything is broken. Need both a data migration and changes to dependent modules.
+/// `{ check_struct_and_function_linking: true, check_struct_layout: true, check_friend_linking: true }`: fully backward compatible
+/// `{ check_struct_and_function_linking: true, check_struct_layout: true, check_friend_linking: false }`: Backward compatible, exclude the friend module declare and friend functions
+/// `{ check_struct_and_function_linking: false, check_struct_layout: true, check_friend_linking: false }`: Dependent modules that reference functions or types in this module may not link. However, fixing, recompiling, and redeploying all dependent modules will work--no data migration needed.
```

Comment:
Can you please point out that anything with check_struct_layout == false cannot ensure type safety?

`... -- no data migration needed. NOTE that if this is false, type safety over a series of upgrades cannot be guaranteed.`

Reply:
I add `if check_struct_layout is false` line at the top of the comments.